extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+  {
+    addToKernelManipulationQueue(
+        queryKernel -> {
+          final StageId stageId = queryKernel.getStageId(stageNumber);
+
+          // We need a specially-decorated ObjectMapper to deserialize key statistics.
+          final StageDefinition stageDef = queryKernel.getStageDefinition(stageId);
+          final ObjectMapper mapper = MSQTasks.decorateObjectMapperForKeyCollectorSnapshot(
+              context.jsonMapper(),
+              stageDef.getShuffleSpec().get().getClusterBy(),
+              stageDef.getShuffleSpec().get().doesAggregateByClusterKey()
+          );
+
+          final ClusterByStatisticsSnapshot keyStatistics;
+          try {
+            keyStatistics = mapper.convertValue(keyStatisticsObject, ClusterByStatisticsSnapshot.class);
+          }
+          catch (IllegalArgumentException e) {
+            throw new IAE(
+                e,
+                "Unable to deserialize the key statistic for stage [%s] received from the worker [%d]",
+                stageId,
+                workerNumber
+            );
+          }
+
+          queryKernel.addResultKeyStatisticsForStageAndWorker(stageId, workerNumber, keyStatistics);
+        }
+    );
+  }
+
+  @Override
+  public void workerError(MSQErrorReport errorReport)
+  {
+    if (!workerTaskLauncher.isTaskCanceledByController(errorReport.getTaskId())) {
+      workerErrorRef.compareAndSet(null, errorReport);
+    }
+  }
+
+  /**
+   * This method intakes all the warnings that are generated by the worker. It is the responsibility of the
+   * worker node to ensure that it doesn't spam the controller with unneseccary warning stack traces. Currently, that
+   * limiting is implemented in {@link MSQWarningReportLimiterPublisher}
+   */
+  @Override
+  public void workerWarning(List<MSQErrorReport> errorReports)
+  {
+    // This check safeguards that the controller doesn't run out of memory. Workers apply their own limiting to
+    // protect their own memory, and to conserve worker -> controller bandwidth.
+    long numReportsToAddCheck = Math.min(
+        errorReports.size(),
+        Limits.MAX_WORKERS * Limits.MAX_VERBOSE_WARNINGS - workerWarnings.size()
+    );
+    if (numReportsToAddCheck > 0) {
+      synchronized (workerWarnings) {
+        long numReportsToAdd = Math.min(
+            errorReports.size(),
+            Limits.MAX_WORKERS * Limits.MAX_VERBOSE_WARNINGS - workerWarnings.size()
+        );
+        for (int i = 0; i < numReportsToAdd; ++i) {
+          workerWarnings.add(errorReports.get(i));
+        }
+      }
+    }
+  }
+
+  /**
+   * Periodic update of {@link CounterSnapshots} from subtasks.
+   */
+  @Override
+  public void updateCounters(CounterSnapshotsTree snapshotsTree)
+  {
+    taskCountersForLiveReports.putAll(snapshotsTree);
+    Optional<Pair<String, Long>> warningsExceeded =
+        faultsExceededChecker.addFaultsAndCheckIfExceeded(taskCountersForLiveReports);
+
+    if (warningsExceeded.isPresent()) {
```

Comment:
Not obvious what `isPresent` means here. It seems that if the `warningsExceeded` `isPresent` we output the warning, but not otherwise. A comment would go a long way to clear up the mystery.

Reply:
The javadoc to `addFaultsAndCheckIfExceeded` describes what it means. I'll add a comment here, too, though.